nixos/modules/tasks/bcache.nix

```diff
@@ -1,13 +1,23 @@
-{ pkgs, ... }:
+{ config, lib, pkgs, ... }:
 
 {
+  options.boot.initrd.services.bcache.enable = (lib.mkEnableOption "bcache support in the initrd") // {
```

Comment:
I'm not sure how to feel about defaulting this to false. Same with lvm and mdraid. In my view, if we want systemd-stage1 to be the default some day, then it *must* be a turn-key solution with failed assertions for anything that can't be turn-key. So anything that users may have implicitly relied on in the old initrd that we can't detect at eval time, I feel like we *have* to include them by default. I like being able to disable things we don't need though.

Maybe we can solve this with a `boot.initrd.systemd.compatibility` option that's enabled by default to include things like bcache, lvm, and mdraid? If systemd-stage1 ever becomes default, we can put a deprecation warning on it for an extended period of time before finally turning it off by default.

Reply:
> if we want systemd-stage1 to be the default some day, then it must be a zero-churn solution

I'm uncertain if I agree, AFAIK it's okay to cause eval breakage across releases. I think it's better that all users benefit from less bloat than a few users be saved for a couple LOC change.

It's important to have feature parity, but I don't think it's imperative that the transition be completely work free for everyone.

Reply:
@lovesegfault Yes, I suppose "zero-churn" wasn't exactly what I wanted to say. Eval errors due to the change are acceptable. Changes in behavior that come with no warning or error, which can cause stage 1 to fail, are unacceptable if systemd-stage-1 is to become default. We cannot detect at eval time if bcache, LVM, or mdraid are necessary for stage 1, so we cannot issue appropriate warnings/errors if users were relying on the fact that the traditional initrd provided those by default.

Reply:
> Eval errors due to the change are acceptable. Changes in behavior that come with no warning or error, which can cause stage 1 to fail, are unacceptable

We are in agreement :)

Reply:
The reason I wanted to add it into this change was that for beta testers, people will have to manually touch their config anyway and will be instructed on what to do with a (WIP) Discourse post that can already inform them about these changes. When the new stage 1 becomes the default, we will print warnings anyway because we are not 100% compatible (see hibernation as another example) so this new behaviour can be added as well.

Sidenote, I don't think it's too relevant for bcache because I don't think a lot of people use this old technology.

Reply:
`bcache` is not old?

c.f. https://en.wikipedia.org/wiki/Bcache

Reply:
I'm fine with landing this IFF we open a tracking issue on what defaults have changed.

Then we can refer and decide what to do before making the options visible.

Reply:
@ElvishJerricco https://github.com/NixOS/nixpkgs/projects/51#card-81007928

Is this sufficient for you, for now?

Reply:
re bcache. The linked Wikipedia article states:
> Overstreet has since been developing the [file system](https://en.wikipedia.org/wiki/File_system) [bcachefs](https://en.wikipedia.org/wiki/Bcachefs), based on ideas first developed in bcache that he said began "evolving ... into a full blown, general-purpose [POSIX](https://en.wikipedia.org/wiki/POSIX) filesystem".[[11]](https://en.wikipedia.org/wiki/Bcache#cite_note-11) He describes bcache as a "prototype" for the ideas that became bcachefs and intends bcachefs to replace bcache.[[12]](https://en.wikipedia.org/wiki/Bcache#cite_note-12)

Reply:
@lovesegfault I would change that card to also mention lvm and mdraid. I'm just not sure how we could possibly turn those into eval-time warnings. Otherwise, sure, for now.

Reply:
The note has been edited, and a reference to this discussion has been added.

I'll land the PR now :)